File: toolchain_py/myfpga/implementation.py
"""Implement a design.

Convert discrete lookup tables and flip flops to logic cells.

"""

from dataclasses import dataclass

import networkx as nx

from myfpga.synthesis import LookUpTable, FlipFlop, ModulePort, FlipFlopInputPort


@dataclass(frozen=True, eq=True)
class LogicCell:
    lut: LookUpTable
    ff: FlipFlop


@dataclass(frozen=True, eq=True)
class LogicCellInputConnection:
    lut_port: int
    logic_cell: LogicCell


class ImplementationError(RuntimeError):
    # FUTURE: Take reference to object which caused the error for better reporting
    # FUTURE: Report errors using verilog source file and line where possible
    pass


class Implementation:

    # def __init__(self, design, device_config):
    def __init__(self, design):
        self.design = design
        source_graph = self.design.build_graph()
        self.clock_input_port = self._sanity_check_flip_flops(source_graph)
        self.graph = self._create_logic_cells(source_graph)

    @staticmethod
    def _sanity_check_flip_flops(graph):
        """Verify that flip flops are configured as expected.

        At this time only a single clock domain is supported.
        If there is clocked logic in the design, it must come from a
        module input port and all flip flops must use it.
        This module port is returned if found, or None if there is
        no clocked logic in the design.

        """
        clock_input_port = None
        for source, sink, port in graph.edges.data('port'):
            if isinstance(sink, FlipFlop) and port is FlipFlopInputPort.clock:
                # At this time FFs must be clocked by a dedicated clock tree,
                # not from programmable logic.
                if not (isinstance(source, ModulePort) and source.is_input):
                    raise ImplementationError(
                        f'{sink.ff.name} clocked from non-module input {source.name}'
                    )

                if clock_input_port is None:
                    clock_input_port = source
                elif source is not clock_input_port:
                    raise ImplementationError(
                        f'{sink.ff.name} should be clocked by main clock source '
                        f'{clock_input_port.name}, not {source.name}'
                    )
        return clock_input_port

    @classmethod
    def _create_logic_cells(cls, source_graph):
        """Merge LUTs and flip flops in the design into combined logic cells.

        Sometimes this cannot be done, such as when an input feeds directly
        into a flip flop or when the output of a LUT is used before it
        passes into a flip flop. In that case additional logic cells are
        created with passthrough LUTs or which bypass their flip flop as needed.

        """
        graph = nx.DiGraph()

        # When we replace nodes with a logic cell, this is how we will
        # reconstruct the connections in the new graph.
        node_replacements = dict()

        # First pass: find LUTs which feed directly into a single FF
        # We cannot merge a LUT and flip flop if anything other than the
        # FF input is using the LUT output.
        for source, sink, port in list(source_graph.edges.data('port')):
            criteria = (
                isinstance(source, LookUpTable)
                and isinstance(sink, FlipFlop)
                and port is FlipFlopInputPort.data
                and source_graph.out_degree(source) == 1
            )
            if criteria:
                logic_cell = LogicCell(lut=source, ff=sink)
                node_replacements[source] = logic_cell
                node_replacements[sink] = logic_cell
                source_graph.remove_edge(source, sink)

        # Second pass: convert the remaining LUTs and FFs into standlone logic cells.
        for source, sink, port in source_graph.edges.data('port'):
            source = cls._replace_node(source, node_replacements, is_source=True)
            sink = cls._replace_node(sink, node_replacements, is_source=False)

            # TODO: Find a better way of handling this
            # We basically have to redirect from the port of the now-merged
            # flip flop into the passthrough port of the LUT attached to it.
            if port is FlipFlopInputPort.data and isinstance(sink, LogicCell) and sink.lut.name == '!passthrough_lut':
                port = 0

            # Strip off FF input ports since they're not useful anymore.
            if isinstance(port, int):
                graph.add_edge(source, sink, port=port)
            elif port is FlipFlopInputPort.clock:
                graph.add_edge(source, sink, port='clock')
            else:
                graph.add_edge(source, sink)

        return graph

    @classmethod
    def _replace_node(cls, node, node_replacements, *, is_source):
        # Logic cells of combined LUT and FF are already in node_replacements
        if node in node_replacements:
            return node_replacements[node]
        elif isinstance(node, LookUpTable):
            # Otherwise we can create a new logic cell with a bypassed flip flop
            logic_cell = LogicCell(lut=node, ff=None)
            node_replacements[node] = logic_cell
            return logic_cell
        elif isinstance(node, FlipFlop):
            # ... or with a passthrough LUT.
            PASSTHROUGH_CONFIG = 0x1010101010101010  # assume port 0 for the connection
            lut = LookUpTable(name='!passthrough_lut', config=PASSTHROUGH_CONFIG)
            logic_cell = LogicCell(lut=lut, ff=node)
            node_replacements[node] = logic_cell
            return logic_cell
        elif isinstance(node, ModulePort):
            # At this point the design, all ports are outputs.
            # They should map directly to I/O cells as-is.
            assert is_source == node.is_input
            return node
        elif not isinstance(node, LogicCell):
            raise NotImplementedError(node)

File: toolchain/toolchain/src/implementation.rs

use std::collections::HashMap;

use petgraph::graph::DiGraph;

use crate::synthesis::{
    DesignGraph, DesignGraphNode, DesignGraphEdge,
    LookUpTable,LookUpTableMask, LookUpTableInput,
    FlipFlop, FlipFlopInput,
    ModulePort, ModulePortDirection,
};


pub type ImplGraph = DiGraph<ImplGraphNode, ImplGraphEdge>;


#[derive(Debug)]
pub enum ImplGraphNode {
    LogicCell(LogicCell),
    ModulePort(ModulePort),
}


#[derive(Debug, Clone, Copy)]
pub enum ImplGraphEdge {
    LogicCellInput(LookUpTableInput),
    LogicCellClock,
    ModulePortInput,
}

impl ImplGraphEdge {
    fn can_connect_to(self, node: &ImplGraphNode) -> bool {
        match (self, node) {
            (ImplGraphEdge::LogicCellInput(_), ImplGraphNode::LogicCell(_)) => true,
            (ImplGraphEdge::LogicCellClock, ImplGraphNode::LogicCell(_)) => true,
            (ImplGraphEdge::ModulePortInput, ImplGraphNode::ModulePort(ModulePort { direction: ModulePortDirection::Output, .. })) => true,
            _ => false,
        }
    }
}


#[derive(Debug)]
pub enum ImplError {
    /// A flip flop is clocked from a non-module input port.
    FlipFlopClockSource { ff: String, clock_source: String },
    /// Flip flops are clocked from multiple domains.
    MultipleClockDomains { ff: String, expected_clock_source: String, actual_clock_source: String },
}


#[derive(Debug)]
pub struct LogicCell {
    pub lut: LookUpTable,

    // If the flip flop is None, then the logic cell output is fed
    // directly from the LUT output. Otherwise it is buffered through
    // the flip flop.
    pub ff: Option<FlipFlop>
}


fn sanity_check_flip_flops(design_graph: &DesignGraph) -> Result<(), ImplError> {
    // For now, only a single clock domain is supported.
    // Furthermore, the clock must originate from a dedicated external clock port.
    // It is okay for the design to not use a clock though,
    // if the logic function is purely combinational.
    let mut main_clock_port: Option<(&ModulePort, petgraph::graph::NodeIndex)> = None;

    for edge in design_graph.raw_edges() {
        let source = &design_graph[edge.source()];
        let sink = &design_graph[edge.target()];
        // Verify that the sink is a flip flop clock input -- if not, skip
        if let (DesignGraphNode::FlipFlop(ff), DesignGraphEdge::FlipFlopInput(FlipFlopInput::Clock)) = (sink, edge.weight) {
            // Verify that the source is a module port -- if not, error
            if let DesignGraphNode::ModulePort(port @ ModulePort { direction: ModulePortDirection::Input, .. }) = source {
                // Verify that if the main clock (i.e. the first clock) has been
                // found that all other flip flops use the same net for their
                // clock input.
                let port_index = edge.source();
                if let Some((main_clock_port, main_clock_port_id)) = main_clock_port {
                    if port_index != main_clock_port_id {
                        return Err(ImplError::MultipleClockDomains {
                            ff: ff.name.clone(),
                            expected_clock_source: main_clock_port.name.clone(),
                            actual_clock_source: source.name(),
                        });
                    }
                }
                else {
                    main_clock_port = Some((port, port_index));
                }
            }
            else {
                return Err(ImplError::FlipFlopClockSource { ff: ff.name.clone(), clock_source: source.name() });
            }
        }
    }
    Ok(())
}


pub fn implement_design(design_graph: DesignGraph) -> Result<ImplGraph, ImplError> {
    sanity_check_flip_flops(&design_graph)?;

    let mut impl_graph = ImplGraph::new();
    let mut node_replacements = HashMap::new();

    // First pass: find LUTs which feed directly into a single flip flop.
    // We cannot merge a LUT and flip flop if anything other than the single
    // flip flop input is using the LUT output.
    for edge in design_graph.raw_edges() {
        let source_index = edge.source();
        let sink_index = edge.target();

        let out_degree = design_graph.edges(source_index).count();
        if out_degree != 1 { continue; }

        if let (
            DesignGraphNode::LookUpTable(lut),
            DesignGraphNode::FlipFlop(ff),
            DesignGraphEdge::FlipFlopInput(FlipFlopInput::Data),
        ) = (&design_graph[source_index], &design_graph[sink_index], edge.weight) {
            let logic_cell = LogicCell {
                lut: lut.clone(),
                ff: Some(ff.clone())
            };
            let node = impl_graph.add_node(ImplGraphNode::LogicCell(logic_cell));
            node_replacements.insert(source_index, node);
            node_replacements.insert(sink_index, node);
        }
    }

    // TODO: Find a cleaner way to implement this. I would have liked to use
    //   a closure, but the borrow checker would not allow it because of the
    //   mutable references.
    use petgraph::graph::NodeIndex;
    fn replace_node(node_index: NodeIndex, design_graph: &DesignGraph, impl_graph: &mut ImplGraph, node_replacements: &mut HashMap<NodeIndex, NodeIndex>) -> NodeIndex {
        // If the node already exists in the replacement map, use its replacement.
        if let Some(replacement_node_index) = node_replacements.get(&node_index) {
            *replacement_node_index
        }
        else {
            // Otherwise we have to create a replacement node.
            let impl_node = match &design_graph[node_index] {
                DesignGraphNode::LookUpTable(lut) => {
                    ImplGraphNode::LogicCell(LogicCell { lut: lut.clone(), ff: None })
                },
                DesignGraphNode::FlipFlop(ff) => {
                    let lut = LookUpTable {
                        name: String::from("$lut$passthrough"),
                        // assume Port A for the passthrough connection
                        mask: LookUpTableMask::new(0b1010_1010_1010_1010)
                    };
                    ImplGraphNode::LogicCell(LogicCell { lut, ff: Some(ff.clone()) })
                },
                DesignGraphNode::ModulePort(port) => ImplGraphNode::ModulePort(port.clone())
            };
            // Insert the replacement node into the new graph and the
            // replacement map for later reference.
            let impl_node = impl_graph.add_node(impl_node);
            node_replacements.insert(node_index, impl_node);
            impl_node
        }
    };

    // Second pass: convert the remaining LUTs and flip flops into
    // standalone logic cells.
    for edge in design_graph.raw_edges() {
        let source_index = replace_node(edge.source(), &design_graph, &mut impl_graph, &mut node_replacements);
        let sink_index = replace_node(edge.target(), &design_graph, &mut impl_graph, &mut node_replacements);

        // Skip edges between the LUT and flip flop in a newly merged logic cell.
        if source_index == sink_index {
            if let DesignGraphEdge::FlipFlopInput(ff_input) = edge.weight {
                assert!(matches!(ff_input, FlipFlopInput::Data));
                continue;
            }
        }

        let impl_edge = match edge.weight {
            DesignGraphEdge::LookUpTableInput(lut_input) => ImplGraphEdge::LogicCellInput(lut_input),
            DesignGraphEdge::ModulePortInput => ImplGraphEdge::ModulePortInput,
            DesignGraphEdge::FlipFlopInput(FlipFlopInput::Clock) => ImplGraphEdge::LogicCellClock,

            // Connections intended for a flip flop should be redirected through the
            // passthrough LUT in its logic cell.
            DesignGraphEdge::FlipFlopInput(FlipFlopInput::Data) => ImplGraphEdge::LogicCellInput(LookUpTableInput::A),
        };
        impl_graph.add_edge(source_index, sink_index, impl_edge);

        // println!("{:?} ({:?}) ->\n  {:?} ({:?})\n  via {:?}\n\n", &impl_graph[source_index], source_index, &impl_graph[sink_index], sink_index, impl_edge);
    }

    // Verify that the generated edges are correct for the node
    // they connect to.
    // FUTURE: Can I use the type system to make the compiler
    //   do this check for me? I don't know that I can with a
    //   graph library expecting homogenous node types and
    //   arbitrary weight types.
    for edge in impl_graph.raw_edges() {
        let target = &impl_graph[edge.target()];
        assert!(
            edge.weight.can_connect_to(target),
            "Cannot connect {:?} to {:?}", edge.weight, target,
        );
    }
    Ok(impl_graph)
}
